capsules/src/ieee802154/xmac.rs

```diff
@@ -0,0 +1,632 @@
+//! wrapping an actual radio driver with the same interface, and can be used
+//! as the backend for a `capsules::ieee802154::mac::MacDevice`. Given a radio
+//! driver `RF233Device` and `kernel::hil::time::Alarm` and
+//! `kernel::hil::rng::RNG` implementations, the necessary modifications to a
+//! board configuration are shown below (for `imix`):
+//!
+//! ```rust
+//! use capsules::xmac::XMac;
+//! type XMacDevice = capsules::xmac::XMac<'static, RF233Device, Alarm>;
+//! // ...
+//! // XMac needs two buffers in addition to those provided to the RF233 driver.
+//! //   1. stores actual packet contents to free the SPI buffers used by the
+//! //      radio for transmitting preamble packets
+//! //   2. buffer to receive packets into before passing them up the stack
+//! static mut XMAC_BUF: [u8; radio::MAX_BUF_SIZE] = [0x00; radio::MAX_BUF_SIZE];
+//! static mut XMAC_RX_BUF: [u8; radio::MAX_BUF_SIZE] = [0x00; radio::MAX_BUF_SIZE];
+//! // ...
+//! let xmac: &XMacDevice = static_init!(XMacDevice, XMac::new(rf233, alarm, rng));
+//! rng.set_client(xmac);
+//!
+//! // Hook up the radio to the XMAC implementation. We can now treat `xmac` as
+//! // the radio driver.
+//! rf233.set_transmit_client(xmac);
+//! rf233.set_receive_client(xmac, &mut RF233_RX_BUF);
+//! rf233.set_power_client(xmac);
+//!
+//! xmac.initialize(&mut RF233_BUF, &mut RF233_REG_WRITE, &mut RF233_REG_READ);
+//! xmac.initialize_xmac(&mut XMAC_BUF);
+//!
+//! // We can now use the XMac driver to instantiate a MacDevice
+//! let xmac_mac = static_init!(
+//!     capsules::ieee802154::mac::MacDevice<'static, XMacDevice>,
+//!     capsules::ieee802154::mac::MacDevice::new(xmac));
+//! xmac.set_transmit_client(xmac_mac);
+//! xmac.set_receive_client(xmac_mac, &mut XMAC_RX_BUF);
+//! xmac.set_config_client(xmac_mac);
+//! ```
+//
+// Author: Jean-Luc Watson
+// Date: Nov 21 2017
+//
+
+use core::cell::Cell;
+use ieee802154::mac::Mac;
+use kernel::ReturnCode;
+use kernel::common::take_cell::TakeCell;
+use kernel::hil::radio;
+use kernel::hil::rng::{self, RNG};
+use kernel::hil::time::{self, Alarm, Frequency, Time};
+use net::ieee802154::*;
+
+// Time the radio will remain awake listening for packets before sleeping.
+// Observing the RF233, receive callbacks for preambles are generated only after
+// having been awake for more than 4-6 ms; 10 ms is a safe amount of time where
+// we are very likely to pick up any incoming preambles, and is half as much
+// as the 20 ms lower bound in Buettner et al.
+const WAKE_TIME_MS: u32 = 10;
+// Time the radio will sleep between wakes. Configurable to any desired value
+// less than or equal to the max time the transmitter sends preambles before
+// abandoning the transmission.
+const SLEEP_TIME_MS: u32 = 250;
+// Time the radio will continue to send preamble packets before aborting the
+// transmission and returning ENOACK. Should be at least as large as the maximum
+// sleep time for any node in the network.
+const PREAMBLE_TX_MS: u32 = 250;
```

Comment:
Seems like this should probably be 251, though it would very rarely matter

Reply:
Nice catch - agreed that it will probably never matter, but I've updated it to 251 just in case.